### ISSUE TITLE
viewVars from view "bleeding" into element

### ISSUE THREAD

**Issue by advitum** [cakephp/cakephp#12782]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Application Skeleton Version: 3.6.5
* CakePHP Version: 3.6.13
* Platform and Target: Apache (any?); no Database; Request via Http

### What you did
I set a view var in the controller:

```
$this->set('varA', 'testA');
```

The corresponding view (in full):

```
<?php

$varB = 'testB';
echo $this->element('test-element', [
    'varC' => 'testC'
]);

?>
```

And the Element "test-element" in full:

```
<?php

debug($varA);
debug($varB);
debug($varC);

?>
```

### What happened
The three debug-calls in the element reveal:

```
$varA === 'testA'
$varB is undefined
$varC === 'testC'
```

### What you expected to happen
```
$varA should be undefined
$varB is undefined
$varC === 'testC'
```

Obviously, `$varC` can be accessed because we explicitly passed it when calling `element()`. And obviously `$varB` is undefined since is only was defined in the view and not passed to the element.

However, I expected `$varB` to be undefined also, since we did not explicitly pass it to `element()`. It seems like all view vars defined in the controller via `$this->set()` are automatically available in every element included in the view. 

Is this a feature or a bug? Aren't elements supposed to be separate from the view (and its view vars) to be more flexible?

**Comment by advitum** [comment]
I did, thanks. Might be wise to add that to the ISSUE_TEMPLATE, since that only asks for "CakePHP Application Skeleton Version".

**Comment by ADmad** [comment]
All variable set through controller are available in elements too. This is a feature not a bug.

**Comment by markstory** [comment]
If you want an isolated rendering context for an element you can use a [cell](https://book.cakephp.org/3.0/en/views/cells.html) instead.

**Comment by advitum** [comment]
Another example (not that much of a security concern, but an actual example from a recent project):

In my element used to render status messages:

```
switch($type ?? null) {
	case 'error':
		$icon = 'frown-open';
		$class = 'error';
		break;
	case 'success':
		$icon = 'smile-beam';
		$class = 'success';
		break;
	default:
		$icon = 'person-sign';
		break;
}
```

I later output that class (without `h()`, I know where that variable was set…) right into the HTML tag.

It's not very hard to imagine passing a view var named "class" from the controller to the view:

```
$this->set('class', $classMyUserJustEnteredIntoAForm);
```

Obviously, I will escape that in the view before outputting it, who knows where that variable has been… But that same value will be sent to the element where it is output unescaped.

Much harder to catch, much harder to test for, since it only will occur when in the right controller action, after entering a class into the form, and only if the `$type` in the element is neither "error" nor "success". And even if I happen to do that: Since the offending value is only output into the `class` attribute of the tag, I don't even see it when testing, only if also happens to break that HTML tag (which I am not even testing for).

**Comment by dereuromark** [comment]
I can see your points, but at the same time, the so far clear concept of view var inheritance was also quite well working out.

I recommend you make a PR for the docs to clarify the current behavior as this is the indended way of them working with passed down (from controller) as well as passed in (from template) data.

**Comment by dereuromark** [comment]
Maybe there could be a opt-in config option to isolate, an `'isolated' => true/false` or sth for the element() call for when needed.

**Comment by markstory** [comment]
@dereuromark I don't think we need another mode for element. Folks who want separate rendering contexts should use cells.

**Comment by advitum** [comment]
I just sent a pull request for the docs, this clears things up a bit, I think. In my opinion, some kind of isolation mode would be great. Without that, one has to be careful with variables with unexpected origin.

Maybe an alternative to an isolation mode would be making the view vars which were defined in the call to `element()` available in another place, too. Something like this:

```
$this->element('account', [
	'user' => $user
]);
```

In the element:

```
echo $user; // This variable might still come from the broader view var scope.

echo $this->elementVars['user']; // This variable can only come from the second parameter of `element()`.
```

In other words: The vars array passed to `elements()` is made available in the element as is, which is a reliable way for the element to prevent scope bleeding.

That would also be backwards compatible.

**Comment by markstory** [comment]
Variables explicitly passed into an element will overwrite the values defined elsewhere in the view. I don't like the idea of having another way to reference element view variables as it adds more complexity with little benefit. Elements have always inherited the view context as it matches how includes work in PHP, twig, and other templating systems like jinja, ERB behave.

I can understand how this behavior not being explicitly documented may have been surprising. If you need isolated view rendering contexts you can use cells, or subclass `View` in your application and change how `element()` works.

**Comment by advitum** [comment]
The changes to the docs mitigate the pitfalls a bit. Just another thing to keep in mind. Works for me. Thanks for listening!
